keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,544 @@
+  fields being accientally included in the apply configuration resulting
+  in fields being accidentally set to incorrect values and/or fields accidentally
+  being accidentally being clamed as owned.
+
+Both sig-api-machinery and wg-api-expression agree that this enhancement is
+required before server side apply to be promoted to GA.
+
+### Goals
+
+Introduce a typesafe, programmatic way to call server side apply using the typed
+client in client-go.
+
+Express all apply configurations in Go that can be expressed in
+YAML. Specifically, an apply request must include only field that are set by
+applier and exclude those not set by applier.
+
+Validate this enhancement meets the needs of developers:
+
+- An developer not directly involved in this enhancement successfully converts
+  a 1st party controller (one in github.com/kubernetes/kubernetes) to us this
+  enhancement.
+- A representative group of the developer community is made aware of this
+  proposed enhancement, is given early access to it via a fork of
+  controller-tools with the requisite generators, and is given the opportunity
+  to try it out and provide feedback.
+
+### Non-Goals
+
+Enhancements to client-go's dynamic client. The client-go dynamic client already
+supports Apply via Patch, which is adequate for the dynamic client.
+
+Protobuf support. Apply does not support protobuf, and it will not be added with
+this enhancement.
+
+## Proposal
+
+`Apply` functions should be included in the typed clients generated for
+client-go and should accept the apply configuration using a strongly typed
+representation, which will need to be generated for this purpose.
+
+### Risks and Mitigations
+
+#### Poor adoption
+
+Risk: Developers adoption is poor, either because the asthetics/ergonomics are
+not to their liking or the functinality is insufficient to do what they need to
+do with it. This could lead to (a) poor server side apply adoption, and/or (b)
+developers building alternate solutions.
+
+Mitigation: We are working with the kubebuilder community to
+get hands on feedback from developers to guide our design decisions around
+asthetics/egronomics with a goal of having both client-go and kubebuilder
+take an aligned approach to adding apply to clients in a typesafe way.
+
+## Design Details
+
+### Apply functions
+
+The client-go typed clients will be extended to include Apply functions, e.g.:
+
+```go
+func (c *deployments) Apply(ctx Context, deployment *appsv1apply.Deployment, fieldManager string, metav1.ApplyOptions) (*Deployment, error)
+```
+
+`ApplyOptions` will be added to metav1 even though `PatchOptions` will continue
+to be used over the wire. This will make it obvious in the Apply function
+signature that `fieldManager` is required.
+
+```go
+type ApplyOptions struct {
+  DryRun []string `json:"dryRun,omitempty" protobuf:"bytes,1,rep,name=dryRun"`
+  Force *bool `json:"force,omitempty" protobuf:"varint,2,opt,name=force"`
```

Comment:
Are you sure you shouldn't make this a parameter too? User-like things want false and controller-like things want true, I don't think there's a great default?

Reply:
That's a really good point.

I've added an `UNRESOLVED` to this section for the `Force` field and for how we organize the params for Apply, since I feel like those both need more consideration.